functions.Rmd

```diff
@@ -16,26 +16,36 @@ permutation()
 
 ## Long lines
 
-If a function definition runs over multiple lines, indent the second line to 
-where the definition starts.
+If the function name and definition can't fit on a single line, place each argument on its own double indented line. The trailing `)` and leading `{` should go on the same line as the last argument.
 
 ```{r, eval = FALSE}
 # Good
+long_function_name <- function(
+    a = "a long argument",
+    b = "another argument",
+    c = "another long argument") {
+  # As usual code is indented by two spaces.
+}
+
+# No longer recommended because it doesn't work very well when the function
+# name is very long, which is not uncommon with S3 methods.
 long_function_name <- function(a = "a long argument",
                                b = "another argument",
                                c = "another long argument") {
-  # As usual code is indented by two spaces.
+  ...
 }
 
 # Bad
 long_function_name <- function(a = "a long argument",
   b = "another argument",
   c = "another long argument") {
   # Here it's hard to spot where the definition ends and the
-  # code begins
+  # code begins, and to see all three function arguments
 }
 ```
 
+If a function argument can't fit on a single line, this is a sign you should extract out the long default argument into its own helper function.
```

Comment:
An example here might be helpful -- I often see something like:

```
foo <- function(
  x, y,
  valid = c(
    'option_1',
    'option_2',
    'option_3',
    'option_4',
    'option_5',
    'option_6',
    'option_7',
    'option_8'
  )) {
  # ... body
}
```

IIUC this statement is recommending a refactor:

```
get_default_valid_foo <- function() {
  c(
    'option_1',
    'option_2',
    'option_3',
    'option_4',
    'option_5',
    'option_6',
    'option_7',
    'option_8'
  )
}
foo <- function(x, y, valid = get_default_valid_foo()) {
  # ... body
}
```

is that the correct understanding?

Reply:
Right. I've written it up a bit at https://design.tidyverse.org/def-short.html